src/node_api.cc

```diff
@@ -0,0 +1,2478 @@
+
+  *result = info->ArgsLength();
+  return napi_ok;
+}
+
+napi_status napi_is_construct_call(napi_env env,
+                                   napi_callback_info cbinfo,
+                                   bool* result) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because no V8 APIs are called.
+  CHECK_ARG(result);
+
+  v8impl::CallbackWrapper* info =
+      reinterpret_cast<v8impl::CallbackWrapper*>(cbinfo);
+
+  *result = info->IsConstructCall();
+  return napi_ok;
+}
+
+// copy encoded arguments into provided buffer or return direct pointer to
+// encoded arguments array?
+napi_status napi_get_cb_args(napi_env env,
+                             napi_callback_info cbinfo,
+                             napi_value* buf,
+                             size_t bufsize) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because no V8 APIs are called.
+  CHECK_ARG(buf);
+
+  v8impl::CallbackWrapper* info =
+      reinterpret_cast<v8impl::CallbackWrapper*>(cbinfo);
+
+  info->Args(buf, bufsize);
+  return napi_ok;
+}
+
+napi_status napi_get_cb_this(napi_env env,
+                             napi_callback_info cbinfo,
+                             napi_value* result) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because no V8 APIs are called.
+  CHECK_ARG(result);
+
+  v8impl::CallbackWrapper* info =
+      reinterpret_cast<v8impl::CallbackWrapper*>(cbinfo);
+
+  *result = info->This();
+  return napi_ok;
+}
+
+napi_status napi_get_cb_data(napi_env env,
+                             napi_callback_info cbinfo,
+                             void** result) {
+  // Omit NAPI_PREAMBLE and GET_RETURN_STATUS because no V8 APIs are called.
+  CHECK_ARG(result);
+
+  v8impl::CallbackWrapper* info =
+      reinterpret_cast<v8impl::CallbackWrapper*>(cbinfo);
+
+  *result = info->Data();
+  return napi_ok;
+}
+
+napi_status napi_call_function(napi_env env,
+                               napi_value recv,
+                               napi_value func,
+                               size_t argc,
+                               const napi_value* argv,
+                               napi_value* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(result);
+
+  std::vector<v8::Local<v8::Value>> args(argc);
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(env);
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+
+  v8::Local<v8::Value> v8recv = v8impl::V8LocalValueFromJsValue(recv);
+
+  for (size_t i = 0; i < argc; i++) {
+    args[i] = v8impl::V8LocalValueFromJsValue(argv[i]);
+  }
+
+  v8::Local<v8::Value> v8value = v8impl::V8LocalValueFromJsValue(func);
+  RETURN_STATUS_IF_FALSE(v8value->IsFunction(), napi_invalid_arg);
+
+  v8::Local<v8::Function> v8func = v8value.As<v8::Function>();
+  auto maybe = v8func->Call(context, v8recv, argc, args.data());
+
+  if (try_catch.HasCaught()) {
+    return napi_set_last_error(napi_pending_exception);
+  } else {
+    CHECK_MAYBE_EMPTY(maybe, napi_generic_failure);
```

Comment:
I'm not sure, but I guess if the arguments to `Call()` were invalid in some way so that the function could not even be called? It seems safest to check it anyway.